drivers/FastCRC.cpp

```diff
@@ -0,0 +1,184 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "platform/mbed_assert.h"
+#include "FastCRC.h"
+#include <stdio.h>
+
+namespace mbed {
+
+FastCRC::FastCRC(crc_polynomial_type_t polynomial) :
+                 _polynomial_type(polynomial), _inital_value(0x0),
+                 _final_xor(0x0), _reflect_data(false), _reflect_remainder(false)
+{
+    // Set the non-default polynomial specific parameters
+    if (CRC_16BIT_CCITT == _polynomial_type) {
+        _inital_value = ~(0x0);
+    } else if (CRC_16BIT_IBM == _polynomial_type) {
+        _reflect_data = true;
+        _reflect_remainder = true;
+    } else if (CRC_32BIT == _polynomial_type) {
+        _inital_value = ~(0x0);
+        _final_xor  = ~(0x0);
+        _reflect_data = true;
+        _reflect_remainder = true;
+    }
+}
+
+FastCRC::~FastCRC() {
+    deinit();
+}
+
+int32_t FastCRC::init(void)
```

Comment:
What is the intention with init/deinit methods, instead of doing this in ctor/dtor (dtor does call deinit, but init is postponed).

Reply:
Yes init is not mandatory for all techniques, we create table during init and allocate memory, which may result into error.
Also, some hardware CRCs require special init.